### ISSUE TITLE
Default to `no_std` in all but `bitcoin` crate.

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#2413]
Explicit trumps implicit any day, we should use the explicit 

ref: https://github.com/rust-bitcoin/rust-bitcoin/pull/2408#issuecomment-1913265945

**Comment by Kixunil** [comment]
Why except `bitcoin`?

**Comment by tcharding** [comment]
Oh, I got mixed up with no-std and requiring an allocator, you are correct, including bitcoin.

**Comment by krisbitney** [comment]
This may be slightly off-topic, but I'm wondering how close rust-bitcoin is to full support in a no_std environment like Wasm? Is it currently possible to sign a taproot transaction with a local private key in no_std mode?

**Comment by apoelstra** [comment]
I believe the answer is "yes" as far as no-std goes, but we may have WASM-specific problems with rust-secp256k1 because it links to a C library.

**Comment by krisbitney** [comment]
Amazing, thank you. I know that Substrate uses the secp256k1 rust library in a Wasm environment. At least, I saw it on their dependency list.

They also use this crate, which they wrote for no_std:
https://crates.io/crates/libsecp256k1

I'm not sure of the specifics. I'm just doing research. I'll have to do some testing. Thank you again!

**Comment by Kixunil** [comment]
> This may be slightly off-topic, but I'm wondering how close rust-bitcoin is to full support in a no_std environment like Wasm?

FTR you don't actually need `no_std` for wasm to work. I made a production-ready piece that also performs taproot signing using this crate with `std` enabled. :)

**Comment by krisbitney** [comment]
> > This may be slightly off-topic, but I'm wondering how close rust-bitcoin is to full support in a no_std environment like Wasm?
> 
> FTR you don't actually need `no_std` for wasm to work. I made a production-ready piece that also performs taproot signing using this crate with `std` enabled. :)

Amazing! Thank you.

**Comment by dpc** [comment]
> Do we want to support that, though? WASM+std is some crazy half-working Frankenstein monster where everything compiles but you call the wrong function (mostly SystemTime) and suddenly you panic. It's not really something practical to support IMO without building a custom static checking framework.

@TheBlueMatt : We're using in Fedimint and it works good enough. System time is PITA, but we have semgrep lints, wasm unit, integration and e2e tests, and recently  we've figured a way to "grep" the wasm binary and flag banned function call, which should be fast and reliable way. (`twiggy paths --regex wasm_bg.wasm 'std::time::Instant::now.*'`).

So all in all, I think it's quite workable.

**Comment by Kixunil** [comment]
In my case the only reason I use `std` is because I don't want to bother with secp256k1 contexts. `bitcoin` doesn't call any such functions and the only way it can do IO is via IO traits (I use byte slices/vecs).

But I'd absolutely love to get rid of it for the peace of mind and potential space optimization. If we can get `secp256k1` have static contexts without `std` that'd be great.

> only asking if we want to officially support that in rust-bitcoin or not

I think we can and should at least until `no_std` is more ergonomic. Calling any system function (other than alloc) should be immediately obvious anyway because there shouldn't be a reason to need it. The check @dpc suggests sounds great and we should add it to CI.

**Comment by tcharding** [comment]
We want this in leaf crates in 1.0, right?

**Comment by Kixunil** [comment]
The change is backwards-compatible as long as features are pre-defined correctly.

**Comment by apoelstra** [comment]
> We want this in leaf crates in 1.0, right?

Yes. Though possibly there will be some case-by-case exceptions.

**Comment by tcharding** [comment]
Only `bitcoin` and `io` remain to be done. Removing the 1.0

**Comment by mpbagot** [comment]
Both `bitcoin` and `io` have feature gated `no_std`:
`#![cfg_attr(not(feature = "std"), no_std)]` in `io`
and
`#![cfg_attr(all(not(feature = "std"), not(test)), no_std)]` in `bitcoin`

Is that sufficient?

**Comment by apoelstra** [comment]
Interesting. Seems like we're really close to being able to just blanket-`#[no_std]` the whole crate then?

**Comment by mpbagot** [comment]
All of the other crates include a snippet like:
```rust
#[cfg(feature = "std")]
extern crate std;
```

I might be misunderstanding something, but isn't a blanket `#[no_std]` followed by that above snippet functionally equivalent to:
```rust
#![cfg_attr(not(feature = "std"), no_std)]
```
?

**Comment by apoelstra** [comment]
I *think* it is, in which case we should just copy the other crates' style into bitcoin and io and we'd have completed this issue.

**Comment by tcharding** [comment]
Yes they are functionally the same (caveat my comment below). The argument is that explicit trumps implicit: https://github.com/rust-bitcoin/rust-bitcoin/issues/2413

And also that its mildly easier to reason about if we have more simple attributes.

**Comment by tcharding** [comment]
> `#![cfg_attr(not(feature = "std"), no_std)] `

I _think_ that is is not sufficient because IIUC there can be some missing test coverage because `cargo test` messes with std-ness. I'm being explicitly vague because I don't know exactly. Anyways, a plain `#[no_std]` saves me/us having to remember exactly what `cargo test` does ...
